File: player-relations/force_directed.js
var color = d3.scale.category20();
var w = 1200, h = 800;
var s = $('#teams');
var selected_team = teams[0];

$(function(){
    for (idx in teams) {
        s.append($('<option id="t'+idx+'">'+teams[idx]+'</option>'));
    }
    s.change(function(evt){
        selected_team = $(this).val();
        showGraph(graphs[selected_team]);
    });
    showGraph(graphs[selected_team]);
});

function showGraph(graph) {
    $('#vis').empty();
    var vis = d3.select('#vis').append('svg:svg')
        .attr('width', w)
        .attr('height', h);

    var force = null;
    force = self.force = d3.layout.force()
        .nodes(graph.nodes)
        .links(graph.links)
        .gravity(.04)
        .distance(120)
        .charge(-150)
        .size([w, h])
        .start();

    var link = vis.selectAll('line.link')
        .data(graph.links)
        .enter().append('svg:line')
        .attr('class', 'link')
        .attr('x1', function(d) { return d.source.x; })
        .attr('y1', function(d) { return d.source.y; })
        .attr('x2', function(d) { return d.target.x; })
        .attr('y2', function(d) { return d.target.y; });

    var node = vis.selectAll('g.node')
        .data(graph.nodes)
      .enter().append('svg:g')
        .attr('class', 'node')
        .call(force.drag);

    node.append('circle')
        .attr('class', 'node')
        .attr('r', 7)
        .style('fill', function(d) { return color(d.group); });

    node.append('svg:text')
        .attr('class', function(d) { return (selected_team === d.team) ? 'nodetext current' : 'nodetext'; })
        .attr('dx', 12)
        .attr('dy', '.35em')
        .text(function(d) { return d.name; });

    force.on('tick', function() {
      link.attr('x1', function(d) { return d.source.x; })
          .attr('y1', function(d) { return d.source.y; })
          .attr('x2', function(d) { return d.target.x; })
          .attr('y2', function(d) { return d.target.y; });

      node.attr('transform', function(d) { return 'translate(' + d.x + ',' + d.y + ')'; });
    });
}

File: player-relations/force_directed.py
# -*- coding: utf-8 -*-
# Built a graph structure for each participating country showing players'
# connections to team (country) and club mates. Players in the country graphs 
# are grouped by club.
#
# Force Layout
# http://bl.ocks.org/950642
# https://github.com/mbostock/d3/wiki/Force-Layout

import csv, json, itertools
from collections import defaultdict

graphs = {}
players = {}
players_by_teams = defaultdict(list)
players_by_clubs = defaultdict(list)
clubs = []
teams = []

def get_id(li, item):
    if item not in li:
        li.append(item)
    return li.index(item)


def add_links(iterable, item):
    for i in iterable:
        players[i]['nodeid'] = get_id(nodes,
            {'name': players[i]['name'], 'group': get_id(clubs, players[i]['club']), 'team': players[i]['team']})
        # don't link player to himself
        if players[item]['nodeid'] != players[i]['nodeid']:
            links.append({'source': players[item]['nodeid'], 'target': players[i]['nodeid']})


fcsv = open('players.csv', 'r')
reader = csv.reader(fcsv)
headers = reader.next()
for idx, record in enumerate(reader):
    firstname, lastname, age, height, team, position, appearances, goals, club = record
    # with team and club name without country suffix so player name is unique
    player = ('%s %s (%s) %s' % (firstname, lastname, team, ' '.join(club.split(' ')[:-1]))).strip()
    players[player] = {'idx': idx, 'name': player, 'team': team, 'club': club}
    players_by_clubs[club].append(player)
    players_by_teams[team].append(player)
fcsv.close()

for team, pls in players_by_teams.items():
    nodes = []
    links = []
    teams.append(team)
    for p in pls:
        team = players[p]['team']
        club = players[p]['club']
        name = players[p]['name']
        players[p]['nodeid'] = get_id(nodes,
            {'name': name, 'group': get_id(clubs, club), 'team': team})
        add_links(players_by_clubs[club], name)
    graphs[team] = {'nodes': nodes, 'links': links}

js = 'var teams=%s;var graphs=%s;' % (json.dumps(sorted(teams)), json.dumps(graphs))
print( js )

